### PR TITLE
Handful of cards - Set 4

### PR DESCRIPTION
**Pull request by Twet1**
image
image
image
image
image
image
image

### CODE REVIEW

File: server/game/cards/04_JTL/events/KoiogranTurn.ts [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+            id: '9695562265',
+            internalName: 'koiogran-turn',
+        };
+    }
+
+    public override setupCardAbilities() {
+        this.setEventAbility({
+            title: 'Ready a Fighter or Transport unit with 6 or less power',
+            targetResolver: {
+                cardCondition: (card) => card.isUnit() && card.getPower() <= 6 && (card.hasSomeTrait(Trait.Fighter) || card.hasSomeTrait(Trait.Transport)),
```
Comment (AMMayberry1) [inline server/game/cards/04_JTL/events/KoiogranTurn.ts:17]: nit: `hasSomeTrait()` can accept a list of traits directly. Same for Mist Hunter

File: test/server/cards/04_JTL/events/KoiogranTurn.spec.ts [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+                phase: 'action',
+                player1: {
+                    hand: ['koiogran-turn'],
+                    groundArena: ['atst'],
+                    spaceArena: [{ card: 'padawan-starfighter', exhausted: true }, 'trade-federation-shuttle', { card: 'first-light#headquarters-of-the-crimson-dawn', damage: 3 }]
+                },
+                player2: {
+                    hand: ['superlaser-blast'],
+                    spaceArena: ['imperial-interceptor'],
+
```
Comment (AMMayberry1) [inline test/server/cards/04_JTL/events/KoiogranTurn.spec.ts:15]: nit: empty space

File: test/server/cards/04_JTL/events/KoiogranTurn.spec.ts [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+            contextRef.setupTest({
+                phase: 'action',
+                player1: {
+                    hand: ['koiogran-turn'],
+                    groundArena: ['atst'],
+                    spaceArena: [{ card: 'padawan-starfighter', exhausted: true }, 'trade-federation-shuttle', { card: 'first-light#headquarters-of-the-crimson-dawn', damage: 3 }]
+                },
+                player2: {
+                    hand: ['superlaser-blast'],
+                    spaceArena: ['imperial-interceptor'],
```
Comment (AMMayberry1) [inline test/server/cards/04_JTL/events/KoiogranTurn.spec.ts:14]: Can we add something with exactly 6 power to check the filter

File: test/server/cards/04_JTL/events/KoiogranTurn.spec.ts [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+            expect(context.player1).toBeAbleToSelectExactly([context.padawanStarfighter, context.tradeFederationShuttle, context.imperialInterceptor]);
+            context.player1.clickCard(context.padawanStarfighter);
+            expect(context.padawanStarfighter.exhausted).toBeFalse();
+
+            // Defeat all units
+            context.player2.clickCard(context.superlaserBlast);
+            context.koiogranTurn.moveTo('hand');
+
+            // Card played but not triggered
+            context.player1.clickCard(context.koiogranTurn);
```
Comment (AMMayberry1) [inline test/server/cards/04_JTL/events/KoiogranTurn.spec.ts:32]: We can skip tests like this in the future, we have a lot of tests confirming that events with simple targeting just fizzle when played with no target

File: test/server/cards/04_JTL/events/TandemAssault.spec.ts [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+            context.player1.clickCard(context.tielnFighter);
+            context.player1.clickCard(context.resupplyCarrier);
+
+            // Remove space unit for next test
+            context.player2.clickCard(context.takedown);
+            context.player2.clickCard(context.padawanStarfighter);
+
+            context.tandemAssault.moveTo('hand');
+
+            // Tandem Assault played, no viable targets
```
Comment (AMMayberry1) [inline test/server/cards/04_JTL/events/TandemAssault.spec.ts:46]: In this case these tests are good since it's a chained effect and those can be a little subtle. For this test though, could we put out a ground unit that is legal to attack with in order to confirm that attacking with a space unit first is required? Could just play a ground unit from hand and move to next phase. Also fine to create a new test setup since it's kind of a different situation

File: server/game/cards/04_JTL/units/DeathSpaceSkirmisher.ts [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+    protected override getImplementationId () {
+        return {
+            id: '1303370295',
+            internalName: 'death-space-skirmisher'
+        };
+    }
+
+    public override setupCardAbilities () {
+        this.addWhenPlayedAbility({
+            title: 'If you control another space unit, you may exhaust a unit.',
```
Comment (AMMayberry1) [inline server/game/cards/04_JTL/units/DeathSpaceSkirmisher.ts:15]: nit: don't need the "if" or "you may" clause on a triggered ability. Should just be "exhaust a unit"

File: test/server/cards/04_JTL/units/DeathSpaceSkirmisher.spec.ts [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+            });
+
+            const { context } = contextRef;
+
+            // Play Death Space Skirmisher, and exhaust a unit
+            context.player1.clickCard(context.deathSpaceSkirmisher);
+            expect(context.player1).toHavePassAbilityButton();
+            expect(context.player1).toBeAbleToSelectExactly([context.kyloRen, context.kylosTieSilencer, context.battlefieldMarine, context.greenSquadronAwing, context.deathSpaceSkirmisher]);
+
+
```
Comment (AMMayberry1) [inline test/server/cards/04_JTL/units/DeathSpaceSkirmisher.spec.ts:25]: nit: empty space

File: test/server/cards/04_JTL/units/MassassiTacticalOfficer.spec.ts [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+            context.player1.clickCard(context.massassiTacticalOfficer);
+            context.player1.clickPrompt('Attack with a Fighter unit. It get +2/+0 for this attack');
+
+            expect(context.player1).toBeAbleToSelectExactly([context.wingLeader, context.starWingScout]);
+
+            // Attack with Wing Leader, observe buff on attack only
+            context.player1.clickCard(context.wingLeader);
+            context.player1.clickCard(context.p2Base);
+            expect(context.p2Base.damage).toBe(4);
+            expect(context.wingLeader.getPower()).toBe(2);
```
Comment (AMMayberry1) [inline test/server/cards/04_JTL/units/MassassiTacticalOfficer.spec.ts:23]: Need to check that Tactical Officer is exhausted

File: server/game/cards/04_JTL/units/MistHunterTheFindsmansPursuit.ts [hunk 9]
```diff
@@ -0,0 +1,37 @@
...
+        return this.cardsPlayedThisPhaseWatcher.someCardPlayed((entry) => entry.playedBy === controller && entry.card.hasSomeTrait(Trait.BountyHunter) || entry.card.hasSomeTrait(Trait.Pilot));
+    }
+
+    public override setupCardAbilities () {
+        this.addOnAttackAbility({
+            title: 'Draw a card',
+            optional: true,
+            immediateEffect: AbilityHelper.immediateEffects.conditional({
+                condition: (context) => this.hasPlayedASpecificTraitThisPhase(context.source.controller),
+                onTrue: AbilityHelper.immediateEffects.draw((context) => ({ target: context.source.controller })),
```
Comment (AMMayberry1) [inline server/game/cards/04_JTL/units/MistHunterTheFindsmansPursuit.ts:32]: nit: we're switching over to using `context.player` instead of `context.source.controller`, it behaves more correctly in edge cases. There are a few other uses of it in these cards that need to be converted as well

File: test/server/cards/04_JTL/units/MistHunterTheFindsmansPursuit.spec.ts [hunk 10]
```diff
@@ -0,0 +1,45 @@
...
+                player1: {
+                    hand: ['boshek#charismatic-smuggler', 'ketsu-onyo#old-friend', 'battlefield-marine'],
+                    spaceArena: ['mist-hunter#the-findsmans-pursuit']
+                },
+            });
+
+            const { context } = contextRef;
+
+            // Not triggered, due to no card played with correct Trait
+            context.player1.clickCard(context.battlefieldMarine);
```
Comment (AMMayberry1) [inline test/server/cards/04_JTL/units/MistHunterTheFindsmansPursuit.spec.ts:15]: Let's also play a Bounty Hunter / Pilot card from the opponent to make sure that's ignored by the filter

File: server/game/cards/04_JTL/units/DeathSpaceSkirmisher.ts [hunk 11]
```diff
@@ -0,0 +1,32 @@
...
+
+    public override setupCardAbilities () {
+        this.addWhenPlayedAbility({
+            title: 'Exhaust a unit.',
+            optional: true,
+            targetResolver: {
+                cardTypeFilter: WildcardCardType.Unit,
+                immediateEffect: AbilityHelper.immediateEffects.conditional({
+                    condition: (context) => {
+                        const spaceUnits = context.source.controller.getUnitsInPlay(ZoneName.SpaceArena)
```
Comment (AMMayberry1) [inline server/game/cards/04_JTL/units/DeathSpaceSkirmisher.ts:21]: Needs to be switched to context.player

**Review by AMMayberry1 — APPROVED** [review]
